hack/test-go.sh

```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env bash
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# This script runs go test.
+#
+# Usage:
+#   hack/test-go.sh [OPTIONS]
+# Options:
+#   -failfast:    Do not start new tests after the first test failure.
+#                 Different from `failfast` flag in `go test`, it will stop all the tests.
+#
+# Examples:
+#   hack/test-go.sh
+#   hack/test-go.sh -failfast
+
+REPO_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+cd $REPO_ROOT
+
+pid=$$
+failfast=
+
+for o in "$@"; do
+  case $o in
+  -failfast|--failfast)
+    failfast=true
+    ;;
+  esac
+done
+
+go_test() {
+  pkgs=$1
+  go test -race -v "${pkgs}" \
+    | grep -v '\[no test files\]$'\
+    | (
+        while read -r line; do
+          echo "$line"
+          if [[ "$failfast" = "true" && "$line" = "--- FAIL: "* ]]; then
```

Comment:
Personally, I don't like the way analysis tests output.

Reply:
The output might be changed over Go versions.

Reply:
well, I will research other solutions.

Reply:
How about this tool: https://github.com/gotestyourself/gotestsum.
# Principle
`gotestsum` runs `go test -json`. `go test` will output json log. The json log is like
```json
{"Time":"2022-07-04T11:12:04.226882+08:00","Action":"output","Package":"github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1","Test":"TestString","Output":"--- FAIL: TestString (0.00s)\n"}
{"Time":"2022-07-04T11:12:04.226889+08:00","Action":"output","Package":"github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1","Test":"TestString/%s_pointer_test","Output":"    --- FAIL: TestString/%s_pointer_test (0.00s)\n"}
{"Time":"2022-07-04T11:12:04.226895+08:00","Action":"fail","Package":"github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1","Test":"TestString/%s_pointer_test","Elapsed":0}
{"Time":"2022-07-04T11:12:04.226899+08:00","Action":"fail","Package":"github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1","Test":"TestString","Elapsed":0}
{"Time":"2022-07-04T11:12:04.226901+08:00","Action":"output","Package":"github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1","Output":"FAIL\n"}
{"Time":"2022-07-04T11:12:04.227604+08:00","Action":"output","Package":"github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1","Output":"FAIL\tgithub.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1\t0.669s\n"}
{"Time":"2022-07-04T11:12:04.228643+08:00","Action":"fail","Package":"github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1","Elapsed":0.67}
{"Time":"2022-07-04T11:12:04.229741+08:00","Action":"output","Package":"github.com/karmada-io/karmada/pkg/apis/config/v1alpha1","Output":"?   \tgithub.com/karmada-io/karmada/pkg/apis/config/v1alpha1\t[no test files]\n"}
{"Time":"2022-07-04T11:12:04.229758+08:00","Action":"skip","Package":"github.com/karmada-io/karmada/pkg/apis/config/v1alpha1","Elapsed":0}
{"Time":"2022-07-04T11:12:04.229769+08:00","Action":"output","Package":"github.com/karmada-io/karmada/pkg/apis/networking/v1alpha1","Output":"?   \tgithub.com/karmada-io/karmada/pkg/apis/networking/v1alpha1\t[no test files]\n"}
```
Then `gotestsum` watches and parses the json event from `go test` output. If `--max-fails` flag is set in `gotestsum`, it will cancel the `go test` process when test failures reaches.

# Test
```
$ gotestsum --max-fails=1 -- -race -failfast -v ./pkg/...
✓  pkg/apis/cluster/validation (cached)
✓  pkg/apis/work/v1alpha2 (cached)
∅  pkg/aggregatedapiserver
∅  pkg/apis/cluster
∅  pkg/apis/cluster/install
✓  pkg/controllers/context (cached)
✓  pkg/descheduler/core (cached)
✓  pkg/detector (cached)
✓  pkg/controllers/status (cached)
✓  pkg/descheduler (cached)
✓  pkg/printers (cached)
✓  pkg/karmadactl/cmdinit/cert (cached)
✓  pkg/printers/internalversion (cached)
✓  pkg/registry/search/storage (cached)
✓  pkg/resourceinterpreter/defaultinterpreter (cached)
✓  pkg/karmadactl (cached)
✓  pkg/estimator/server (cached)
✓  pkg/scheduler/core/spreadconstraint (cached)
✓  pkg/util/imageparser (cached)
✓  pkg/util/informermanager/keys (cached)
✓  pkg/scheduler/core (cached)
✓  pkg/util/interpreter (cached)
✓  pkg/util/names (cached)
✓  pkg/util (cached)
✓  pkg/util/lifted (cached)
✓  pkg/scheduler (cached)
✓  pkg/version (cached)
✓  pkg/util/helper (cached)
✓  pkg/webhook/configuration (cached)
✓  pkg/util/validation (cached)
✓  pkg/webhook/federatedresourcequota (cached)
✓  pkg/util/overridemanager (cached)

=== Failed
=== FAIL: pkg/apis/cluster/v1alpha1 TestString/%s_pointer_test (0.00s)
    cluster_helper_test.go:46: %s pointer test String() = cluster1, want cluster1
    --- FAIL: TestString/%s_pointer_test (0.00s)

=== FAIL: pkg/apis/cluster/v1alpha1 TestString (unknown)

=== Errors
DONE 443 tests, 2 failures, 23 errors in 3.045s
ERROR ending test run because max failures was reached
```

Reply:
I never heard about this tool before, so I don't know if it is mature enough. 
I just tried on my side under `karmada-io/karmada` root path:
```
# go install gotest.tools/gotestsum@latest
go: downloading gotest.tools/gotestsum v1.8.1
go: downloading github.com/dnephin/pflag v1.0.7
go: downloading golang.org/x/sys v0.0.0-20210615035016-665e8c7367d1
go: downloading golang.org/x/term v0.0.0-20201126162022-7de9c90e9dd1

# gotestsum --max-fails=1 -- -race -failfast -v ./pkg/...
∅  pkg/aggregatedapiserver
∅  pkg/apis/cluster
∅  pkg/apis/cluster/install
✓  pkg/apis/cluster/v1alpha1 (49ms)
✓  pkg/apis/work/v1alpha2 (56ms)
✓  pkg/apis/cluster/validation (101ms)
...

=== Failed
=== FAIL: pkg/karmadactl/cmdinit/utils TestDownloadFile (unknown)
Downloading...[ 100.00% ]
Downloading...[ 100.00% ]
Download complete.--- PASS: TestDownloadFile (1.26s)

DONE 447 tests, 1 failure in 36.290s
ERROR ending test run because max failures was reached
```
I didn't get much useful information to figure out the reason of the failure.

Actually, the test is passed and works as expected:
```bash
# go test ./pkg/karmadactl/cmdinit/utils/... -run=TestDownloadFile -count=50
ok  	github.com/karmada-io/karmada/pkg/karmadactl/cmdinit/utils	13.752s

```

Reply:
Can we introduce flag or environment to `make test`, such as with `-failfast` the tests will be run by sequence, otherwise run by pararenal?

Reply:
I may misunderstand the purpose of `failfast`: accelerating go test, or easy to locate the fail test?

Reply:
> easy to locate the fail test

+1

Reply:
I don't think it is difficult to locate the fail test. Just search `-- FAIL` in the web
<img width="1410" alt="image" src="https://user-images.githubusercontent.com/17907604/177295904-a267d21b-afe8-4d83-819b-cd3020cd7e07.png">

Personally, I think failfast is  unfriendly. Imagine developer push code with two error tests(e.g. Test1 and Test2). Then CI reports Test 1 failed. Then he correct Test1 and repush. Then CI tells Test2 failed.

IMO, CI shall reports all the errors.

Reply:
ok. we can defer this for a while.
Then close both this PR and issue?

Reply:
> Then close both this PR and issue?

Yes, it's meaningless recently.

/close